### PR TITLE
Add an alternative Karras et al. stochastic scheduler for VE models 

### PR DESCRIPTION
**Pull request by anton-l**
The required number of inference steps went down from 2000 to 50 with comparable quality :tada: 

The algorithm is slightly modified to rely on pre- and post-processing steps integrated into the VE UNet's forward pass (centering and scaling by sigma), so it's specific to VE only.

### CODE REVIEW

File: src/diffusers/pipelines/stochatic_karras_ve/pipeline_stochastic_karras.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+from ...schedulers import KarrasVeScheduler
+
+
+class KarrasVePipeline(DiffusionPipeline):
+    """
+    Stochastic sampling from Karras et al. [1] tailored to the Variance-Expanding (VE) models [2].
+    Use Algorithm 2 and the VE column of Table 1 from [1] for reference.
+
+    [1] Karras, Tero, et al. "Elucidating the Design Space of Diffusion-Based Generative Models." arXiv preprint arXiv:2206.00364 (2022).
+    [2] Song, Yang, et al. "Score-based generative modeling through stochastic differential equations." In Proc. ICLR (2021).
```
Comment (patil-suraj) [inline src/diffusers/pipelines/stochatic_karras_ve/pipeline_stochastic_karras.py:17]: (nit) could we also put a arxiv link ?

File: src/diffusers/pipelines/stochatic_karras_ve/pipeline_stochastic_karras.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+        for i in tqdm(self.scheduler.timesteps):
+            # here sigma_t == t_i from the paper
+            # sigmas are multiplied by 2 for compatibility with the original VE scheduler's sigma range
+            sigma = self.scheduler.schedule[i]
+            sigma_prev = self.scheduler.schedule[i - 1] if i > 0 else 0
+
+            # 1. Select temporarily increased noise level sigma_hat
+            # 2. Add new noise to move from sample_i to sample_hat
+            sample_hat, sigma_hat = self.scheduler.get_model_inputs(sample, sigma, generator=generator)
```
Comment (patil-suraj) [inline src/diffusers/pipelines/stochatic_karras_ve/pipeline_stochastic_karras.py:52]: `get_model_inputs` doesn't seem to describe the function well. Maybe let's call it `add_noise_to_input` or `churn_input`.

File: src/diffusers/pipelines/stochatic_karras_ve/pipeline_stochastic_karras.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+                step_output = self.scheduler.step_correct(
+                    model_output,
+                    sigma_hat,
+                    sigma_prev,
+                    sample_hat,
+                    step_output["prev_sample"],
+                    step_output["derivative"],
+                )
+            sample = step_output["prev_sample"]
+            print(i, sample.min().item(), sample.max().item())
```
Comment (patil-suraj) [inline src/diffusers/pipelines/stochatic_karras_ve/pipeline_stochastic_karras.py:75]:
```suggestion
```

File: src/diffusers/schedulers/scheduling_karras_ve.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    """
+
+    @register_to_config
+    def __init__(
+        self,
+        sigma_min=0.02,
+        sigma_max=100,
+        s_churn=80,
+        s_min=0.05,
+        s_max=50,
```
Comment (patil-suraj) [inline src/diffusers/schedulers/scheduling_karras_ve.py:42]: (nit) maybe rename the `s_` parameters to `scale_` for example `scale_churn`.

File: src/diffusers/schedulers/scheduling_karras_ve.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+        return sample_hat, sigma_hat
+
+    def step(
+        self,
+        model_output: Union[torch.FloatTensor, np.ndarray],
+        sigma_hat: float,
+        sigma_prev: float,
+        sample_hat: Union[torch.FloatTensor, np.ndarray],
+    ):
+        D = sample_hat + sigma_hat * model_output
```
Comment (patil-suraj) [inline src/diffusers/schedulers/scheduling_karras_ve.py:88]: Let's avoid single letter variable, this `D` also refers to `derivative` no ?

File: src/diffusers/schedulers/scheduling_karras_ve.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    def step_correct(
+        self,
+        model_output: Union[torch.FloatTensor, np.ndarray],
+        sigma_hat: float,
+        sigma_prev: float,
+        sample_hat: Union[torch.FloatTensor, np.ndarray],
+        sample_prev: Union[torch.FloatTensor, np.ndarray],
+        derivative: Union[torch.FloatTensor, np.ndarray],
+    ):
+        D_prev = sample_prev + sigma_prev * model_output
```
Comment (patil-suraj) [inline src/diffusers/schedulers/scheduling_karras_ve.py:103]: same comment as above.

File: tests/test_modeling_utils.py [hunk 7]
```diff
@@ -920,3 +922,19 @@ def test_ddpm_ddim_equality_batched(self):
...
+    def test_karras_ve_pipeline(self):
+        model_id = "google/ncsnpp-celebahq-256"
+        model = UNet2DModel.from_pretrained(model_id)
+        scheduler = KarrasVeScheduler(tensor_format="pt")
+
+        pipe = KarrasVePipeline(unet=model, scheduler=scheduler)
+
+        generator = torch.manual_seed(0)
+        image = pipe(num_inference_steps=20, generator=generator, output_type="numpy")["sample"]
+
```
Comment (patil-suraj) [inline tests/test_modeling_utils.py:936]: Nice test!

**Review by patil-suraj — APPROVED** [review]
Great work @anton-l ! It looks good to me, just left some nits.

 A more general comment. This `scheduler` goes a bit against our design of single `step` function. But given that the scheduler requires two model evaluations and the second eval depends on the output of first we can't really have a single `step` function. We also need to do some bookkeeping outside the scheduler like storing `sigma` , `sigma_prev` etc. (Looking at the code think we can probably avoid it). cc @patrickvonplaten 

Thanks a lot for working on this!

File: src/diffusers/schedulers/scheduling_utils.py [hunk 8]
```diff
@@ -54,6 +54,9 @@ def log(self, tensor):
 
         raise ValueError(f"`self.tensor_format`: {self.tensor_format} is not valid.")
 
+    def sqrt(self, tensor):
+        return tensor**0.5
```
Comment (patil-suraj) [inline src/diffusers/schedulers/scheduling_utils.py:58]: Let's avoid one line functions :)

File: src/diffusers/schedulers/scheduling_karras_ve.py (hunk 4 above)
Comment (patrickvonplaten) [reply src/diffusers/schedulers/scheduling_karras_ve.py:42]: +1

File: src/diffusers/pipelines/stochatic_karras_ve/pipeline_stochastic_karras.py (hunk 2 above)
Comment (patrickvonplaten) [reply src/diffusers/pipelines/stochatic_karras_ve/pipeline_stochastic_karras.py:52]: I would just call it `get_sigmas(...)`, but I understand where `get_model_inputs` is coming from (to make the API as generic for all schedulers as possible which is also a good argument)

Here I would be in favor of being a bit more specific with something like `get_sigmas(...)` which is something most continuous schedulers have and which could become a common API across continuous schedulers. 
So I think here it makes sense to favor intuitive, readable code over easy-to-use. 

Also it scares me a bit to see that `sample` is an input to the function, it gives the impression that it's used to compute the sigmas. Can we instead maybe just pass `shape` and `device`?

File: src/diffusers/schedulers/scheduling_karras_ve.py (hunk 5 above)
Comment (patrickvonplaten) [reply src/diffusers/schedulers/scheduling_karras_ve.py:88]: +1

File: tests/test_modeling_utils.py [hunk 9]
```diff
@@ -920,3 +922,19 @@ def test_ddpm_ddim_equality_batched(self):
 
         # the values aren't exactly equal, but the images look the same visually
         assert np.abs(ddpm_images - ddim_images).max() < 1e-1
+
+    @slow
+    def test_karras_ve_pipeline(self):
```
Comment (patrickvonplaten) [inline tests/test_modeling_utils.py:927]: Very cool!!! (Did you also try it with the 1024 one?)

File: src/diffusers/schedulers/scheduling_utils.py (hunk 8 above)
Comment (patrickvonplaten) [reply src/diffusers/schedulers/scheduling_utils.py:58]: +1 also `** 0.5` works for all frameworks

File: src/diffusers/schedulers/scheduling_karras_ve.py (hunk 4 above)
Comment (anton-l) [reply src/diffusers/schedulers/scheduling_karras_ve.py:42]: the S likely refers to "stochastic" in the paper (page 7, "Practical considerations" https://arxiv.org/pdf/2206.00364.pdf), so I'm not sure this would make it more explicit. I'll move these to the `pipeline.__call__()` and document them instead.

**Comment by patrickvonplaten** [comment]
> Great work @anton-l ! It looks good to me, just left some nits.
> 
> A more general comment. This `scheduler` goes a bit against our design of single `step` function. But given that the scheduler requires two model evaluations and the second eval depends on the output of first we can't really have a single `step` function. We also need to do some bookkeeping outside the scheduler like storing `sigma` , `sigma_prev` etc. (Looking at the code think we can probably avoid it). cc @patrickvonplaten
> 
> Thanks a lot for working on this!

That's indeed a very important question and I don't really know the best answer here.

Note that we could make it work by making `step` stateful (like we did in the `pndm` scheduler), but then it becomes maybe a bit too magic. 

Also `correct` is a pretty good name here IMO opinion because it "corrects" the initial prediction with a (second) derivative computation.

So @anton-l and @patil-suraj feel free to go ahead with whatever you think is best here!

**Review by patrickvonplaten — APPROVED** [review]
Amazing work Anton! This is the by far best continuous scheduler that we have now and makes the `ve` models usable (maybe even on CPU!)

Also it's great that the code is so clean and includes links to the paper!

Left some remarks, mostly nits good to go for me!

Let's make sure to advertise this well worth some nice examples (maybe linking the NVIDIA author as well)

File: src/diffusers/schedulers/scheduling_karras_ve.py (hunk 4 above)
Comment (patrickvonplaten) [reply src/diffusers/schedulers/scheduling_karras_ve.py:42]: Ok for me

File: src/diffusers/pipelines/stochatic_karras_ve/pipeline_stochastic_karras.py (hunk 2 above)
Comment (anton-l) [reply src/diffusers/pipelines/stochatic_karras_ve/pipeline_stochastic_karras.py:52]: Note that `sample` is also used in computation: `sample_hat = sample + ((sigma_hat**2 - sigma**2)**0.5 * eps)`

Comment (patil-suraj) [reply src/diffusers/pipelines/stochatic_karras_ve/pipeline_stochastic_karras.py:52]: `get_sigmas` sound good to me, but still not clear enough, as it adds noise to the current sigmas rather than giving completly new `sigma`.

> Also it scares me a bit to see that sample is an input to the function, it gives the impression that it's used to compute the sigmas. Can we instead maybe just pass shape and device?

Good poin! In this case we will need to compute the `sample_hat` in the pipeline then. Also, the function then will have to return the `eps` as well which is needed to compute `sample_hat`.  I'm still in favor of `add_noise_to_input` as it makes clear that the function adds noise to the inputs rather using `sample` to compute `sigma`. That's also how the paper paper describes this part.

Comment (anton-l) [reply src/diffusers/pipelines/stochatic_karras_ve/pipeline_stochastic_karras.py:52]: Opting for `add_noise_to_input` for now :)

Comment (patrickvonplaten) [reply src/diffusers/pipelines/stochatic_karras_ve/pipeline_stochastic_karras.py:52]: ah sorry @anton-l you're right - missed that part. Then think we should more or less leave as is and maybe rename to `compute_sigmas(...)` to make sure reader knows that sigmas are dependent on the sample and completed (not taken from a predefined list)

Feel free to go ahead as you want though @anton-l

File: src/diffusers/schedulers/scheduling_karras_ve.py (hunk 4 above)
Comment (patil-suraj) [reply src/diffusers/schedulers/scheduling_karras_ve.py:42]: looking at the function these values are used to scale the stochastic noise so `scale` would be good IMO. 

Also these arguments are a bit low level to expose them them in the `__call __. So maybe keep in scheduler. Having scheduler self-contained would be better IMO.
Or do you think there is some advantage having them in `pipeline` ?

File: tests/test_modeling_utils.py (hunk 9 above)
Comment (anton-l) [reply tests/test_modeling_utils.py:927]: Yes, the results are pretty ok :) 
(Although they can be made better with a bit of grid search to find the optimal `s_churn`, `s_noise`, etc, since the paper was dealing with much smaller models)

File: src/diffusers/schedulers/scheduling_karras_ve.py (hunk 4 above)
Comment (anton-l) [reply src/diffusers/schedulers/scheduling_karras_ve.py:42]: @patil-suraj on second thought, these parameters are best to have inside the scheduler, since they have to be tuned to each specific model. BTW, take a look at the docstrings so far to see if they would fit `scale_`

File: src/diffusers/schedulers/scheduling_karras_ve.py [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+            sigma_min (`float`): minimum noise magnitude
+            sigma_max (`float`): maximum noise magnitude
+            s_noise (`float`): the amount of additional noise to counteract loss of detail during sampling.
+                A reasonable range is [1.000, 1.011].
+            s_churn (`float`): the parameter controlling the overall amount of stochasticity.
+                A reasonable range is [0, 100].
+            s_min (`float`): the start of the sigma range where we add noise (enable stochasticity)
+                A reasonable range is [0, 10].
+            s_max (`float`): the end of the sigma range where we add noise
+                A reasonable range is [0.2, 80].
```
[the remark this answers is not in the document]
Comment (anton-l) [reply src/diffusers/schedulers/scheduling_karras_ve.py:61]: The thing is that after renaming it's harder to refer to their paper counterparts :( 
Let's keep them as is for now, maybe we'll figure something out while implementing the SD scheduler
